text/3803-enum-repr-no-niches.md

```diff
@@ -0,0 +1,709 @@
+
+This might need some way to suppress the warning on the enum itself (`#[repr(_, niches)]`? `#[exhaustive]`?), since certain enums _are_ valid to use across FFI. Examples of this include C99's `_Bool` if it were an enum, enums marked `__attribute__((enum_extensibility(closed)))`, enums such as `CompassDirection` (there are only ever four cardinal directions) and enums otherwise documented as exhaustive.
+
+## Enums with fields
+[enums-with-fields]: #enums-with-fields
+
+The `no_niches` option apply specifically to the enum discriminant. This could be extended to [`#[repr(_)]` enums with fields](https://doc.rust-lang.org/1.86.0/reference/type-layout.html#reprc-enums-with-fields), and would translate to the `no_niches` option being present on the internal discriminant enum.
+
+Building on the reference link:
+
+```rust
+#[repr(C, no_niches)] // `no_niches` was added
+enum MyEnum {
+    A(u32),
+    B(f32, u64),
+    C { x: u32, y: u8 },
+    D,
+}
+
+// Would have this discriminant enum.
+#[repr(C, no_niches)]
+enum MyEnumDiscriminant { A, B, C, D }
+```
+
+## Zero-variant enums
+
+The reference [states](https://doc.rust-lang.org/1.86.0/reference/type-layout.html#r-layout.repr.primitive.constraint):
+> It is an error for zero-variant enums to have a primitive representation
+
+We could consider relaxing this for `no_niches` enums, since these are actually inhabited:
+
+```rust
+#[repr(u8)] // Disallowed.
+pub enum Never {}
+let never = unsafe { core::mem::transmute::<u8, Never>(42) }; // Unsound.
+
+#[repr(u8, no_niches)] // Could be allowed.
+pub enum Empty {}
+let empty = unsafe { core::mem::transmute::<u8, Empty>(42) }; // Sound.
+```
+
+Though when to use this over `pub type Empty = u8;` or `pub struct Empty(u8);` is yet unclear.
+
+## Further progress stable Rust ABI
+
+Rust does not have a stable ABI, but there are [thoughts](https://faultlore.com/blah/swift-abi/) [on](https://docs.rs/abi_stable) [improving](https://docs.rs/stabby) [that](https://github.com/rust-lang/rfcs/pull/3470), and having a way to define an enum's discriminants as ABI-stable could be a useful stepping stone.
+
+If Rust intends to make big strides here, it might make sense to tailor the syntax of this feature towards that? Perhaps something like an `#[abi_stable]` attribute that would also be usable on `struct`s and `trait`s? But it probably shouldn't block this RFC, as it can always be added later.
+
+## Changing the default
+
+Making `#[repr(C)]` and `#[repr(uXX)]` enums have the `no_niches` modifier by default across an edition boundary could make interoperability with C easier and even further remove the footgun (anecdotally, I've seen several people surprised that `#[repr(C)] enum` isn't a good idea for interfacing with C enums).
+
+Would need some way to re-opt-in to niches, maybe `#[repr(_, niches)]`?
```

Comment:
Honestly, I would much prefer something along the lines of `niches(on)` and `niches(off)`, or just having both `no_niches` and `niches` available when this feature is added, for forward-compatibility. This way, people can explicitly opt into having `niches` without worrying about whether the default is changed in a future edition.

Reply:
I've changed the RFC to use `#[repr(_, open)]`, so I don't think `niches(on)`/`niches(off)` really make sense any more. I've added them as alternative syntaxes though.

Regarding forward-compat, I tend to agree, but I would rather not drag the RFC down with extra features. I've added it as an unresolved question instead, to make sure that we consider it before stabilization.